sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala

```diff
@@ -125,7 +125,8 @@ private[sql] object PruneFileSourcePartitions
         getPartitionKeyFiltersAndDataFilters(scan.sparkSession, v2Relation,
           scan.readPartitionSchema, filters, output)
       // The dataFilters are pushed down only once
-      if (partitionKeyFilters.nonEmpty || (dataFilters.nonEmpty && scan.dataFilters.isEmpty)) {
+      if ((partitionKeyFilters.nonEmpty && dataFilters.isEmpty)
```

Comment:
dump question: how does this guarantee only-once?

Reply:
per our off-line discussion, added comment to explain the `if` condition.